django/db/backends/mysql/operations.py

```diff
@@ -177,13 +177,51 @@ def sql_flush(self, style, tables, sequences, allow_cascade=False):
         # NB: The generated SQL below is specific to MySQL
         # 'TRUNCATE x;', 'TRUNCATE y;', 'TRUNCATE z;'... style SQL statements
         # to clear all tables of all data
+        schema_name = self.connection.settings_dict['NAME']
+
         if tables:
+            with self.connection.cursor() as cursor:
+                cursor.execute("""
+                SELECT table_name, table_rows, auto_increment
+                FROM information_schema.tables
+                WHERE table_schema = %s AND
+                  ((auto_increment IS NOT NULL AND auto_increment > 1) OR table_rows > 0);
+                """, (schema_name,))
+                rows = cursor.fetchall()
+
             sql = ['SET FOREIGN_KEY_CHECKS = 0;']
-            for table in tables:
-                sql.append('%s %s;' % (
-                    style.SQL_KEYWORD('TRUNCATE'),
-                    style.SQL_FIELD(self.quote_name(table)),
-                ))
+
+            for table_name, table_rows, auto_increment in rows:
+                if table_name not in tables:
+                    continue
+
+                if table_rows == 0:
```

Comment:
[This is only an estimate on InnoDB tables](https://dev.mysql.com/doc/refman/5.7/en/tables-table.html) and cannot be relied on when InnoDB is used which the default.

> The number of rows. Some storage engines, such as MyISAM, store the exact count. *For other storage engines, such as InnoDB, this value is an approximation, and may vary from the actual value by as much as 40% to 50%. In such cases, use SELECT COUNT(*) to obtain an accurate count.*

In short that means this value could report 0 while there's actually rows in the table and cause errors similar to the one you are experiencing.

Reply:
Thank you!
Hmm.... If it requires to call `SELECT COUNT(*)`, this patch seems to be slower than the existing implementation.

Reply:
I believe you can just drop the `== 0` case here. Doing `DELETE FROM` on 0 rows should be harmless.

Reply:
Thank you for your advice! I see. I'll update a patch.

Reply:
Thank you! The problem seems to be fixed.